src/codegen/monomorphize.lisp

```diff
@@ -199,103 +199,104 @@ recompilation, and also maintains a stack of uncompiled candidates."
-                                                       :collect (make-node-variable :type ty :value arg))))))))
-                  node))
-
-               ;; Function is under applied
-               ((and (null new-vars)
-                     (< (length candidate-args) (length abstraction-vars)))
-                (let* ((remaining-args (subseq abstraction-vars (length candidate-args)))
-
-                       (node-ty (node-type node))
-
-                       (arg-tys (subseq (tc:function-type-arguments node-ty) (length candidate-args)))
-
-                       (ret-ty (tc:function-return-type node-ty)))
-                  (tc:apply-substitution
-                   subs
-                   (make-node-abstraction
-                    :type (tc:make-function-type* arg-tys ret-ty)
-                    :vars remaining-args
-                    :subexpr (apply-ast-substitution ast-subs (node-abstraction-subexpr node))))))
-
-               ;; Function is applied
-               (t
-                (tc:apply-substitution
-                 subs
-                 (make-node-abstraction
-                  :type (tc:make-function-type*
-                         (reverse arg-tys)
-                         new-type)
-                  :vars (append new-vars retained-args)
-                  :subexpr (apply-ast-substitution ast-subs (node-abstraction-subexpr node)))))))))
-
-    (typecheck-node new-node env)
-
-    new-node))
+         (subexpr (apply-ast-substitution ast-subs (node-abstraction-subexpr node))))
+
+    (setf arg-tys (reverse arg-tys))
+
+    (let ((new-node
+            (cond
+              ;; Candidate is fully applied
+              ((= (length (compile-candidate-args candidate)) (length (node-abstraction-vars node)))
+               (tc:apply-substitution
+                subs
+                (make-node-abstraction
+                 :type (tc:make-function-type* arg-tys new-type)
+                 :vars new-vars
+                 :subexpr subexpr)))
+
+              ;; Overapplication
+              ((> (length (compile-candidate-args candidate)) (length (node-abstraction-vars node)))
+
+               ;; If the function is overapplied then there will be extra arguments to apply
+               (let* ((remaining-args (util:drop (length (node-abstraction-vars node)) (compile-candidate-args candidate)))
+
+                      (num-remaining (length remaining-args))
+
+                      (remaing-names (loop :for arg :in remaining-args
+                                           :collect (gentemp)))
+
+                      (remaining-types (loop :for arg :in remaining-args
+                                             :for ty :in (tc:function-type-arguments new-type)
+                                             :collect ty)))
+
+                 ;; All remaining arguments must be unpropagated
+                 (unless (every (lambda (x) (eq x '@@unpropagated)) remaining-args)
```

Comment:
is it ok that EVERY is true for an empty list?

Reply:
Because of the check in the cond branch the list will never be empty.

Reply:
This seems like a great thing to note in a quick comment.